File: frontend/src/lib/marketplace-pricing.ts
import { apiFetch } from './api';

export type MarketplaceId = 'mercado_livre' | 'shopee';

export interface MarketplaceFees {
  commissionPercent: number;
  paymentFeePercent: number;
  taxPercent: number;
  advertisingPercent: number;
  fixedFee: number;
  shippingCost: number;
}

export interface ManualProduct {
  id: string;
  sku: string;
  name: string;
  productCost: number;
  packagingCost: number;
  operationalCost: number;
  desiredMarginPercent: number;
  promotionalMarginPercent: number;
  currentPrice: number;
}

export type PricingAlert =
  | 'sem_preco'
  | 'abaixo_minimo'
  | 'margem_critica'
  | 'abaixo_recomendado'
  | 'saudavel';

export interface PricingResult {
  marketplace: MarketplaceId;
  minimumPrice: number;
  recommendedPrice: number;
  promotionalPrice: number;
  currentProfit: number | null;
  currentMarginPercent: number | null;
  alert: PricingAlert;
}

export interface PricingComparison {
  mercado_livre: PricingResult;
  shopee: PricingResult;
  lowestRecommendedMarketplace: MarketplaceId;
  recommendedDifference: number;
}

export const emptyFees = (): MarketplaceFees => ({
  commissionPercent: 0,
  paymentFeePercent: 0,
  taxPercent: 0,
  advertisingPercent: 0,
  fixedFee: 0,
  shippingCost: 0,
});

export const createProduct = (sequence = Date.now()): ManualProduct => ({
  id: `${sequence}-${Math.random().toString(36).slice(2, 8)}`,
  sku: `SKU-${String(sequence).slice(-4)}`,
  name: 'Novo produto',
  productCost: 0,
  packagingCost: 0,
  operationalCost: 0,
  desiredMarginPercent: 20,
  promotionalMarginPercent: 10,
  currentPrice: 0,
});

const requireFiniteNonNegative = (value: number, label: string) => {
  if (!Number.isFinite(value) || value < 0) {
    throw new Error(`${label} não pode ser negativo.`);
  }
};

const ceilMoney = (value: number) => Math.ceil((value - Number.EPSILON) * 100) / 100;
const roundMoney = (value: number) => Math.round((value + Number.EPSILON) * 100) / 100;

export function calculatePricing(
  product: ManualProduct,
  marketplace: MarketplaceId,
  fees: MarketplaceFees,
): PricingResult {
  if (!product.sku.trim()) throw new Error('Informe o SKU do produto.');
  if (!product.name.trim()) throw new Error('Informe o nome do produto.');
  [
    [product.productCost, 'Custo do produto'],
    [product.packagingCost, 'Embalagem'],
    [product.operationalCost, 'Outros custos'],
    [product.desiredMarginPercent, 'Margem desejada'],
    [product.promotionalMarginPercent, 'Margem promocional'],
    [product.currentPrice, 'Preço atual'],
    [fees.commissionPercent, 'Comissão'],
    [fees.paymentFeePercent, 'Tarifa de pagamento'],
    [fees.taxPercent, 'Impostos'],
    [fees.advertisingPercent, 'Publicidade'],
    [fees.fixedFee, 'Tarifa fixa'],
    [fees.shippingCost, 'Frete'],
  ].forEach(([value, label]) => requireFiniteNonNegative(value as number, label as string));
  if (product.productCost <= 0) throw new Error('O custo do produto precisa ser maior que zero.');
  if (product.promotionalMarginPercent > product.desiredMarginPercent) {
    throw new Error('A margem promocional não pode superar a margem desejada.');
  }

  const variablePercent = fees.commissionPercent + fees.paymentFeePercent
    + fees.taxPercent + fees.advertisingPercent;
  const fixedCost = product.productCost + product.packagingCost + product.operationalCost
    + fees.fixedFee + fees.shippingCost;
  const target = (margin: number) => {
    const denominator = 1 - variablePercent / 100 - margin / 100;
    if (denominator <= 0) throw new Error('Taxas e margem deixam o preço inviável.');
    return ceilMoney(fixedCost / denominator);
  };
  const minimumPrice = target(0);
  const promotionalPrice = target(product.promotionalMarginPercent);
  const recommendedPrice = target(product.desiredMarginPercent);
  let currentProfit: number | null = null;
  let currentMarginPercent: number | null = null;
  let alert: PricingAlert = 'sem_preco';
  if (product.currentPrice > 0) {
    currentProfit = roundMoney(product.currentPrice * (1 - variablePercent / 100) - fixedCost);
    currentMarginPercent = roundMoney(currentProfit / product.currentPrice * 100);
    if (product.currentPrice < minimumPrice) alert = 'abaixo_minimo';
    else if (currentMarginPercent < product.promotionalMarginPercent) alert = 'margem_critica';
    else if (currentMarginPercent < product.desiredMarginPercent) alert = 'abaixo_recomendado';
    else alert = 'saudavel';
  }
  return {
    marketplace,
    minimumPrice,
    recommendedPrice,
    promotionalPrice,
    currentProfit,
    currentMarginPercent,
    alert,
  };
}

export function compareMarketplaces(
  product: ManualProduct,
  fees: Record<MarketplaceId, MarketplaceFees>,
): PricingComparison {
  const mercadoLivre = calculatePricing(product, 'mercado_livre', fees.mercado_livre);
  const shopee = calculatePricing(product, 'shopee', fees.shopee);
  const lowest = mercadoLivre.recommendedPrice <= shopee.recommendedPrice
    ? 'mercado_livre' : 'shopee';
  return {
    mercado_livre: mercadoLivre,
    shopee,
    lowestRecommendedMarketplace: lowest,
    recommendedDifference: roundMoney(Math.abs(mercadoLivre.recommendedPrice - shopee.recommendedPrice)),
  };
}

function parseDelimited(text: string): string[][] {
  const firstLine = text.split(/\r?\n/, 1)[0] ?? '';
  const delimiter = (firstLine.match(/;/g)?.length ?? 0) > (firstLine.match(/,/g)?.length ?? 0) ? ';' : ',';
  const rows: string[][] = [];
  let row: string[] = [];
  let field = '';
  let quoted = false;
  for (let index = 0; index < text.length; index += 1) {
    const char = text[index];
    if (char === '"') {
      if (quoted && text[index + 1] === '"') { field += '"'; index += 1; }
      else quoted = !quoted;
    } else if (char === delimiter && !quoted) {
      row.push(field); field = '';
    } else if ((char === '\n' || char === '\r') && !quoted) {
      if (char === '\r' && text[index + 1] === '\n') index += 1;
      row.push(field); field = '';
      if (row.some((cell) => cell.trim())) rows.push(row);
      row = [];
    } else field += char;
  }
  row.push(field);
  if (row.some((cell) => cell.trim())) rows.push(row);
  return rows;
}

const normalizeHeader = (value: unknown) => String(value ?? '').trim().toLowerCase()
  .normalize('NFD').replace(/[\u0300-\u036f]/g, '').replace(/[^a-z0-9]+/g, '_').replace(/^_|_$/g, '');

const aliases: Record<string, string[]> = {
  sku: ['sku', 'codigo', 'codigo_sku'],
  name: ['nome', 'produto', 'name', 'titulo'],
  productCost: ['custo', 'custo_produto', 'product_cost', 'productcost'],
  packagingCost: ['embalagem', 'custo_embalagem', 'packaging_cost', 'packagingcost'],
  operationalCost: ['outros_custos', 'custo_operacional', 'operational_cost', 'operationalcost'],
  desiredMarginPercent: ['margem_desejada', 'margem', 'desired_margin_percent', 'desiredmarginpercent'],
  promotionalMarginPercent: ['margem_promocional', 'promotional_margin_percent', 'promotionalmarginpercent'],
  currentPrice: ['preco_atual', 'preco', 'sale_price', 'current_price', 'currentprice'],
};

function localizedNumber(value: unknown, fallback = 0): number {
  if (typeof value === 'number') return Number.isFinite(value) ? value : fallback;
  const raw = String(value ?? '').trim().replace(/R\$|%|\s/g, '');
  if (!raw) return fallback;
  const normalized = raw.includes(',') ? raw.replace(/\./g, '').replace(',', '.') : raw;
  const parsed = Number(normalized);
  return Number.isFinite(parsed) ? parsed : fallback;
}

export function mapImportedRows(rows: Array<Record<string, unknown>>): ManualProduct[] {
  return rows.slice(0, 1000).map((source, index) => {
    const normalized = Object.fromEntries(Object.entries(source).map(([key, value]) => [normalizeHeader(key), value]));
    const valueFor = (field: keyof typeof aliases) => {
      const key = aliases[field].find((candidate) => candidate in normalized);
      return key ? normalized[key] : undefined;
    };
    const sku = String(valueFor('sku') ?? '').trim();
    const name = String(valueFor('name') ?? '').trim();
    if (!sku || !name) throw new Error(`Linha ${index + 2}: informe SKU e nome.`);
    return {
      id: `import-${Date.now()}-${index}`,
      sku,
      name,
      productCost: localizedNumber(valueFor('productCost')),
      packagingCost: localizedNumber(valueFor('packagingCost')),
      operationalCost: localizedNumber(valueFor('operationalCost')),
      desiredMarginPercent: localizedNumber(valueFor('desiredMarginPercent'), 20),
      promotionalMarginPercent: localizedNumber(valueFor('promotionalMarginPercent'), 10),
      currentPrice: localizedNumber(valueFor('currentPrice')),
    };
  });
}

export function parseCsv(text: string): ManualProduct[] {
  const rows = parseDelimited(text.replace(/^\uFEFF/, ''));
  if (rows.length < 2) throw new Error('O CSV precisa ter cabeçalho e ao menos um produto.');
  const headers = rows[0];
  return mapImportedRows(rows.slice(1).map((row) => Object.fromEntries(headers.map((header, index) => [header, row[index] ?? '']))));
}

export async function parseProductFile(file: File): Promise<ManualProduct[]> {
  const extension = file.name.split('.').pop()?.toLowerCase();
  if (extension === 'csv') return parseCsv(await file.text());
  if (extension === 'xlsx') {
    const body = new FormData();
    body.append('file', file);
    const response = await apiFetch('/v1/pricing/import', { method: 'POST', body });
    const payload = await response.json().catch(() => ({}));
    if (!response.ok) {
      throw new Error(payload.detail || payload.error || `Falha ao importar Excel (${response.status}).`);
    }
    if (!Array.isArray(payload.products)) {
      throw new Error('O importador retornou uma resposta inválida.');
    }
    return mapImportedRows(payload.products);
  }
  throw new Error('Formato não suportado. Use CSV ou XLSX.');
}

File: src/openjarvis/zeusex/marketplace_pricing.py
"""Precificação manual e determinística para marketplaces.

O módulo não acessa contas externas nem altera anúncios. As entradas podem vir
da interface manual, de arquivos ou, futuramente, de adaptadores de API.
"""

from __future__ import annotations

from dataclasses import dataclass
from decimal import Decimal, ROUND_CEILING, ROUND_HALF_UP
from typing import Iterable

SUPPORTED_MARKETPLACES = ("mercado_livre", "shopee")
MONEY = Decimal("0.01")
PERCENT = Decimal("0.01")
HUNDRED = Decimal("100")


def _decimal(value: Decimal | int | float | str) -> Decimal:
    try:
        result = Decimal(str(value))
    except Exception as exc:
        raise ValueError(f"Valor numérico inválido: {value!r}.") from exc
    if not result.is_finite():
        raise ValueError("Valores financeiros precisam ser finitos.")
    return result


def _money(value: Decimal) -> Decimal:
    return value.quantize(MONEY, rounding=ROUND_HALF_UP)


def _price(value: Decimal) -> Decimal:
    return value.quantize(MONEY, rounding=ROUND_CEILING)


def _percent(value: Decimal) -> Decimal:
    return value.quantize(PERCENT, rounding=ROUND_HALF_UP)


@dataclass(frozen=True, slots=True)
class MarketplaceFees:
    """Taxas editáveis de um canal, sem valores presumidos."""

    marketplace: str
    commission_percent: Decimal = Decimal("0")
    payment_fee_percent: Decimal = Decimal("0")
    tax_percent: Decimal = Decimal("0")
    advertising_percent: Decimal = Decimal("0")
    fixed_fee: Decimal = Decimal("0")
    shipping_cost: Decimal = Decimal("0")

    def __post_init__(self) -> None:
        marketplace = self.marketplace.strip().lower().replace(" ", "_")
        if marketplace not in SUPPORTED_MARKETPLACES:
            raise ValueError("Marketplace suportado: mercado_livre ou shopee.")
        object.__setattr__(self, "marketplace", marketplace)
        for name in (
            "commission_percent",
            "payment_fee_percent",
            "tax_percent",
            "advertising_percent",
            "fixed_fee",
            "shipping_cost",
        ):
            value = _decimal(getattr(self, name))
            if value < 0:
                raise ValueError(f"{name} não pode ser negativo.")
            object.__setattr__(self, name, value)
        if self.variable_percent >= HUNDRED:
            raise ValueError("A soma das taxas percentuais precisa ser menor que 100%.")

    @property
    def variable_percent(self) -> Decimal:
        return (
            self.commission_percent
            + self.payment_fee_percent
            + self.tax_percent
            + self.advertising_percent
        )


@dataclass(frozen=True, slots=True)
class ManualProduct:
    """Produto cadastrado manualmente ou importado de planilha."""

    sku: str
    name: str
    product_cost: Decimal
    packaging_cost: Decimal = Decimal("0")
    operational_cost: Decimal = Decimal("0")
    desired_margin_percent: Decimal = Decimal("20")
    promotional_margin_percent: Decimal = Decimal("10")
    current_price: Decimal | None = None

    def __post_init__(self) -> None:
        sku = self.sku.strip()
        name = self.name.strip()
        if not sku:
            raise ValueError("O produto precisa de um SKU.")
        if not name:
            raise ValueError("O produto precisa de um nome.")
        object.__setattr__(self, "sku", sku)
        object.__setattr__(self, "name", name)
        for field_name in (
            "product_cost",
            "packaging_cost",
            "operational_cost",
            "desired_margin_percent",
            "promotional_margin_percent",
        ):
            value = _decimal(getattr(self, field_name))
            if value < 0:
                raise ValueError(f"{field_name} não pode ser negativo.")
            object.__setattr__(self, field_name, value)
        if self.current_price is not None:
            current_price = _decimal(self.current_price)
            if current_price < 0:
                raise ValueError("current_price não pode ser negativo.")
            object.__setattr__(self, "current_price", current_price)
        if self.product_cost <= 0:
            raise ValueError("O custo do produto precisa ser maior que zero.")
        if self.desired_margin_percent >= HUNDRED:
            raise ValueError("A margem desejada precisa ser menor que 100%.")
        if self.promotional_margin_percent > self.desired_margin_percent:
            raise ValueError("A margem promocional não pode superar a margem desejada.")

    @property
    def base_cost(self) -> Decimal:
        return self.product_cost + self.packaging_cost + self.operational_cost


@dataclass(frozen=True, slots=True)
class PricingResult:
    product: ManualProduct
    fees: MarketplaceFees
    minimum_price: Decimal
    recommended_price: Decimal
    promotional_price: Decimal
    current_profit: Decimal | None
    current_margin_percent: Decimal | None
    alert: str

    @property
    def marketplace(self) -> str:
        return self.fees.marketplace


def _target_price(base_cost: Decimal, fixed_cost: Decimal, variable: Decimal, margin: Decimal) -> Decimal:
    denominator = Decimal("1") - variable / HUNDRED - margin / HUNDRED
    if denominator <= 0:
        raise ValueError("Taxas e margem deixam o preço matematicamente inviável.")
    return _price((base_cost + fixed_cost) / denominator)


def calculate_pricing(product: ManualProduct, fees: MarketplaceFees) -> PricingResult:
    """Calcula preços e saúde do preço atual para um marketplace."""

    variable = fees.variable_percent
    minimum = _target_price(product.base_cost, fees.fixed_fee + fees.shipping_cost, variable, Decimal("0"))
    recommended = _target_price(
        product.base_cost,
        fees.fixed_fee + fees.shipping_cost,
        variable,
        product.desired_margin_percent,
    )
    promotional = _target_price(
        product.base_cost,
        fees.fixed_fee + fees.shipping_cost,
        variable,
        product.promotional_margin_percent,
    )

    profit: Decimal | None = None
    real_margin: Decimal | None = None
    if product.current_price is None or product.current_price <= 0:
        alert = "sem_preco"
    else:
        current = product.current_price
        profit = current * (Decimal("1") - variable / HUNDRED) - (
            product.base_cost + fees.fixed_fee + fees.shipping_cost
        )
        real_margin = profit / current * HUNDRED
        if current < minimum:
            alert = "abaixo_minimo"
        elif real_margin < product.promotional_margin_percent:
            alert = "margem_critica"
        elif real_margin < product.desired_margin_percent:
            alert = "abaixo_recomendado"
        else:
            alert = "saudavel"

    return PricingResult(
        product=product,
        fees=fees,
        minimum_price=minimum,
        recommended_price=recommended,
        promotional_price=promotional,
        current_profit=_money(profit) if profit is not None else None,
        current_margin_percent=_percent(real_margin) if real_margin is not None else None,
        alert=alert,
    )


@dataclass(frozen=True, slots=True)
class MarketplaceComparison:
    sku: str
    mercado_livre: PricingResult
    shopee: PricingResult
    lowest_recommended_marketplace: str
    recommended_difference: Decimal


def compare_marketplaces(
    product: ManualProduct,
    mercado_livre_fees: MarketplaceFees,
    shopee_fees: MarketplaceFees,
) -> MarketplaceComparison:
    """Compara os dois canais sem fazer qualquer ação externa."""

    mercado_livre = calculate_pricing(product, mercado_livre_fees)
    shopee = calculate_pricing(product, shopee_fees)
    if mercado_livre.recommended_price <= shopee.recommended_price:
        lowest = "mercado_livre"
    else:
        lowest = "shopee"
    difference = abs(mercado_livre.recommended_price - shopee.recommended_price)
    return MarketplaceComparison(
        sku=product.sku,
        mercado_livre=mercado_livre,
        shopee=shopee,
        lowest_recommended_marketplace=lowest,
        recommended_difference=_money(difference),
    )


def calculate_batch(
    products: Iterable[ManualProduct],
    mercado_livre_fees: MarketplaceFees,
    shopee_fees: MarketplaceFees,
    *,
    maximum_products: int = 1000,
) -> list[MarketplaceComparison]:
    """Precifica até mil produtos, preservando a ordem da importação."""

    items = list(products)
    if len(items) > maximum_products:
        raise ValueError(f"O lote aceita no máximo {maximum_products} produtos.")
    seen: set[str] = set()
    for product in items:
        normalized = product.sku.casefold()
        if normalized in seen:
            raise ValueError(f"SKU duplicado no lote: {product.sku}.")
        seen.add(normalized)
    return [
        compare_marketplaces(product, mercado_livre_fees, shopee_fees)
        for product in items
    ]


__all__ = [
    "ManualProduct",
    "MarketplaceComparison",
    "MarketplaceFees",
    "PricingResult",
    "calculate_batch",
    "calculate_pricing",
    "compare_marketplaces",
]
